drivers/gpio/gpio_nrfx.c

```diff
@@ -47,42 +45,39 @@ static inline const struct gpio_nrfx_cfg *get_port_cfg(const struct device *port
 	return port->config;
 }
 
-static int gpiote_channel_alloc(atomic_t *mask, uint32_t abs_pin,
+static int gpiote_channel_alloc(uint32_t abs_pin,
 				nrf_gpiote_polarity_t polarity)
 {
-	for (uint8_t channel = 0; channel < GPIOTE_CH_NUM; ++channel) {
-		atomic_val_t prev = atomic_or(mask, BIT(channel));
-
-		if ((prev & BIT(channel)) == 0) {
-			nrf_gpiote_event_t evt =
-				offsetof(NRF_GPIOTE_Type, EVENTS_IN[channel]);
+	uint8_t channel;
 
-			nrf_gpiote_event_configure(NRF_GPIOTE, channel, abs_pin,
-						   polarity);
-			nrf_gpiote_event_clear(NRF_GPIOTE, evt);
-			nrf_gpiote_event_enable(NRF_GPIOTE, channel);
-			nrf_gpiote_int_enable(NRF_GPIOTE, BIT(channel));
-			return 0;
-		}
+	if (nrfx_gpiote_channel_alloc(&channel) != NRFX_SUCCESS) {
+		return -ENODEV;
```

Comment:
This return value propagates to `gpio_pin_interrupt_configure()` which lists the expected error returns and the conditions that evoke them.  `-ENODEV` is not one of them, and I don't believe it accurately represents the failure mode: all the devices are present, but one doesn't have sufficient resources to satisfy the request.

My proposal is to change this to `-ENOSPC`, then add this to the `gpio.h` description of `gpio_pin_interrupt_configure()`:

    @retval -ENOSPC if the configuration requires additional resources that
                    are not currently available, but may be made available if
                    released by other users.

I'm open to other error values than `-ENOSPC`; that's just the closest to semantically similar I can identify:

    ENOSPC 28 No space left on device

We could touch on that in the API telecon today too @carlescufi

Reply:
You've raised a valid point. `-ENODEV` is not listed as a possible return value of `gpio_pin_interrupt_configure` and it should not be returned here. However, this function's previous implementation was also capable of returning such error code. At the same time, modifying `drivers/gpio.h` seems like a more significant change since it concerns all platforms, not only the `gpio_nrfx.c` implementation. Taking that into account, would it be acceptable for you to keep the behavior of this function (which is a natural continuation of the previous implementation) as it is and discuss & fix the issue you've raised in a separate PR?

Reply:
Yes.  (*sigh*).